sqsworkers/crew.py

```diff
@@ -1,327 +1,225 @@
-                good_workers.append(new_worker)
-        self.crew.logger.info('total number of workers after supervision: %s' % (str(len(good_workers))))
-        self.crew.statsd.gauge('workers.employed', len(good_workers), tags=[self.crew.name])
-        self.crew.workers = good_workers
+
+                messages = self.queue.receive_messages(
+                    AttributeNames=["All"],
+                    MessageAttributeNames=["All"],
+                    MaxNumberOfMessages=self.max_number_of_messages,
+                    WaitTimeSeconds=self.wait_time,
+                )
+
+            if messages:
+
+                metadata = [MessageMetadata(m) for m in messages]
+
+                self.logger.info(
+                    f"processing the following {len(messages)} messages in bulk: {metadata}"
+                )
+
+                task: futures.Future = self._executor.submit(
+                    self.message_processor(messages).start
+                )
+
+                task.add_done_callback(
+                    partial(
+                        self._task_complete,
+                        messages=messages,
+                        metadata=metadata,
+                    )
+                )
+
+                self.statsd.increment(
+                    "process.record.start", len(messages), tags=[]
+                )
+
+            time.sleep(self.polling_interval)
+
+    def _task_complete(
+        self,
+        f: futures.Future,
+        messages: Iterable[Any],
+        metadata=Optional[List[MessageMetadata]],
+    ):
+        """Clean up after task and do any necessary logging."""
+
+        messages = list(messages)
+
+        metadata: List[MessageMetadata] = [
+            MessageMetadata(m) for m in messages
+        ] if metadata is None else metadata
+
+        exception = f.exception()
+
+        if exception is not None:
+
+            self.logger.error(
+                "failed processing {messages} with the following exception: {exception}".format(
+                    exception=repr(exception), messages=metadata
+                ),
+                exc_info=exception,
+            )
+
+            self.statsd.increment(
+                "process.record.failure", len(messages), tags=[]
+            )
+
+        else:
+
+            result = f.result()
+
+            self.logger.info(
+                "successfully processed {} messages: {metadata}".format(
+                    len(result.succeeded),
+                    metadata=[MessageMetadata(m) for m in result.succeeded],
+                )
+            )
+
+            self.statsd.increment(
+                "process.record.success", len(result.succeeded), tags=[]
+            )
+
+            self.logger.info(
+                "failed to process {} messages: {metadata}".format(
+                    len(result.failed),
+                    metadata=[MessageMetadata(m) for m in result.failed],
```

Comment:
Any way that we can add the exception information here like we do in the single message processing `exc_info=exception,` might be hard to add all in one logging message but maybe think of having a few?

Reply:
good point. the individual exceptions with their metadata and tracebacks are [logged by the transformer](https://stash.atlassian.com/projects/DTECH/repos/hydra-transformer/browse/src/msg_processor.py?at=f0fbc870a1a765bbc3e9fa8d4ad5e21a63b7a3e2#199)